File: SortonsEvents/DirectoryViewController.m
//
//  DirectoryViewController.m
//  Belfield
//
//  Created by Brian Henry on 20/04/2015.
//  Copyright (c) 2015 Sortons. All rights reserved.
//

#import "DirectoryViewController.h"
#import "CPDAPIClient.h"
#import "SourcePage.h"
#import "IncludedPageCell.h"
#import "CommonWebViewController.h"
#import <SDWebImage/UIImageView+WebCache.h>
#import "Fomo.h"

@interface DirectoryViewController ()

@property (nonatomic,strong) NSArray *dataSource;
@property (nonatomic,strong) NSMutableArray *filteredData;
@property (strong, nonatomic) IBOutlet UITableView *tableView;

@end

@implementation DirectoryViewController


#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.filteredData.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    
    IncludedPageCell *cell = [tableView dequeueReusableCellWithIdentifier:@"includedPageCell" forIndexPath:indexPath];
    
    cell.cellTitle.text = ((SourcePage *) self.filteredData[indexPath.row]).name;

    cell.detailsLabel.text = @"";
    
    NSString *imageURLString = [NSString stringWithFormat: @"https://graph.facebook.com/%@/picture?type=large", ((SourcePage *) self.filteredData[indexPath.row]).pageId];
    [cell.cellImage sd_setImageWithURL:[NSURL URLWithString:imageURLString]    placeholderImage:nil];
    //    [cell.cellImage sd_setImageWithURL:[NSURL URLWithString:@"https://graph.facebook.com/olivier.poitrey/picture"]
    //                 placeholderImage:[UIImage imageNamed:@"avatar-placeholder.png"]
    //                          options:SDWebImageRefreshCached]; // caches less aggressively

    cell.cellImage.layer.cornerRadius = cell.cellImage.frame.size.width / 2;
    cell.cellImage.clipsToBounds = YES;
    
    // Fix for multiline titles not laying out properly
    [cell layoutIfNeeded];
    
    return cell;
}

#pragma mark - UITableViewDelegate

- (void) tableView: (UITableView *) tableView didSelectRowAtIndexPath: (NSIndexPath *) indexPath {
    
    SourcePage *selectedPage = _filteredData[indexPath.row];
    
    // Get user preference
    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
    BOOL runNative = [defaults boolForKey:@"launch_native_apps_toggle"];
   
    NSString *httpURLString = [NSString stringWithFormat: @"https://facebook.com/%@/", selectedPage.pageId];
    
    if(runNative==TRUE){
        NSString *appURLString = [NSString stringWithFormat: @"fb://profile/%@/", selectedPage.pageId];
        
        NSURL *facebookURL = [NSURL URLWithString:appURLString];
        if ([[UIApplication sharedApplication] canOpenURL:facebookURL]) {
            [[UIApplication sharedApplication] openURL:facebookURL];
        } else {
            [self openInWebView:[NSURL URLWithString:httpURLString]];
        }
    } else {
        [self openInWebView:[NSURL URLWithString:httpURLString]];
    }
 
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}



// duplicate code! (also in EventsViewController)
-(void)openInWebView:(NSURL*)url{
    
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
    UIViewController *myController = [storyboard instantiateViewControllerWithIdentifier:@"CommonWebViewController"];

    CommonWebViewController *cwvs = [CommonWebViewController alloc];
    cwvs = (CommonWebViewController *) myController;
    
    [cwvs setWebViewURL:url];
    
    self.modalTransitionStyle=UIModalTransitionStyleCoverVertical;
    [self presentViewController:myController animated:YES completion:NULL];

}


// Filter the list when searching
- (void)searchBar:(UISearchBar *)searchBar
    textDidChange:(NSString *)searchText {
    
    if(searchText.length==0){
        _filteredData = [NSMutableArray arrayWithArray:_dataSource];
    }else {
        [self.filteredData removeAllObjects];
        for (SourcePage *item in self.dataSource)
        {
            // drop the case for search
            if([item.name.lowercaseString containsString:searchText.lowercaseString]){
                [self.filteredData addObject:item];
            }
        }
    }
    
    [self.tableView reloadData];
    
}


// Hide the keybaord
- (void) searchBarSearchButtonClicked:(UISearchBar *)searchBar {
    [searchBar resignFirstResponder];
}

-(void) scrollViewDidScroll:(UIScrollView *)scrollView{
    // probably goes somewhere else
    scrollView.keyboardDismissMode = UIScrollViewKeyboardDismissModeOnDrag;
}



#pragma mark - UIViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    CPDAPIClient *client = [CPDAPIClient sharedClient];
    
    _dataSource = [client includedPagesFromCache];
    _filteredData = [NSMutableArray arrayWithArray:_dataSource];
    [self.tableView reloadData];
    
    [client getClientPages:[Fomo fomoId]
                   success:^(NSURLSessionDataTask *task, id responseObject) {
                       _dataSource = [CPDAPIClient includedPagesFromJSON:(NSDictionary *)responseObject error:nil];
                       _filteredData = [NSMutableArray arrayWithArray:_dataSource];
                       [self.tableView reloadData];
                   }
                   failure:^(NSURLSessionDataTask *task, NSError *error) {
                       NSLog(@"Failure -- %@", error);
                   }];

}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: SortonsEvents/DirectoryViewController.swift
//
//  DirectoryViewController.swift
//  SortonsEvents
//
//  Created by Brian Henry on 11/10/2016.
//  Copyright © 2016 Sortons. All rights reserved.
//

import UIKit

extension Directory {
    struct Request {}
}

protocol DirectoryViewControllerOutputProtocol {
    func fetchDirectory(_ withRequest: Directory.Request)

    func filterDirectoryTo(_ searchBarInput: String)

    func displaySelectedPageFrom(_ rowNumber: Int)
}

class DirectoryViewController: UIViewController, DirectoryPresenterOutputProtocol {

    @IBOutlet weak var searchBarOutlet: UISearchBar!
    @IBOutlet weak var tableViewOutlet: UITableView!

    var output: DirectoryViewControllerOutputProtocol?
    var data: [Directory.TableViewCellModel]?

    override func viewDidLoad() {
        super.viewDidLoad()

        tableViewOutlet.rowHeight = UITableViewAutomaticDimension
        tableViewOutlet.estimatedRowHeight = 80

        let gestureRecognizer = UIGestureRecognizer()
        gestureRecognizer.cancelsTouchesInView = false
        tableViewOutlet.addGestureRecognizer(gestureRecognizer)

        gestureRecognizer.delegate = self

        let top = UIApplication.shared.statusBarFrame.size.height
        tableViewOutlet.contentInset = UIEdgeInsets(top: top,
                                                   left: 0,
                                                 bottom: 49,
                                                  right: 0)

        tableViewOutlet.setContentOffset(CGPoint(x: 0, y: top * -1), animated: false)

        let request = Directory.Request()
        output?.fetchDirectory(request)
    }

// MARK: DirectoryPresenterOutput

    func presentFetchedDirectory(_ viewModel: Directory.ViewModel) {
        data = viewModel.directory
        tableViewOutlet.reloadData()
    }

    func displayFetchDirectoryFetchError(_ viewModel: Directory.ViewModel) {

    }

    // MARK: UISearchBar
    func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
        output?.filterDirectoryTo(searchText)
    }
}

extension DirectoryViewController: UITableViewDataSource, UITableViewDelegate {
// MARK: - Table view data source
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection: Int) -> Int {
        return data?.count ?? 0
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let sourcePage = data?[indexPath.row] else { return UITableViewCell() }

        let cell = tableView.dequeueReusableCell(withIdentifier: "DirectoryPageCell",
                                                            for: indexPath) as? DirectoryTableViewCell
        cell?.setDirectorySourcePage(sourcePage)

        return cell ?? UITableViewCell()
    }

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let row = indexPath.row
        output?.displaySelectedPageFrom(row)
        tableView.deselectRow(at: indexPath, animated: true)
    }
}

extension DirectoryViewController: UIGestureRecognizerDelegate {
// MARK: UIGestureRecogniserDelegate
    func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer,
                           shouldRecognizeSimultaneouslyWith otherGestureRecognizer: UIGestureRecognizer) -> Bool {
        searchBarOutlet.resignFirstResponder()

        return false
    }
}
